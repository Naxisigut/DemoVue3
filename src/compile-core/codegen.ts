export function generate(ast){
  const code = `return function render(_ctx, _cache, $props, $setup, $data, $options) {
    with (_ctx) {
      return "hi"
    }
  }`

  return {
    code
  }
}