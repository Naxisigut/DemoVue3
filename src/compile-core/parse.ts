import { NodeTypes } from "./ast"

export function baseParse(content: any){
  const context = createParseContext(content)
  return createRoot(parseChildren(context))
}

function createParseContext(content: any) {
  return {
    source: content
  }
}

function createRoot(children: any) {
  return {
    children
  }
}

function parseChildren(context: any) {
  const nodes: any = []
  const node = parseInterpolation(context)
  nodes.push(node)
  return nodes
}

function parseInterpolation(context: any) {
  const openDelimitter = "{{"
  const closeDelimitter = "}}"
  const closeIndex = context.source.indexOf(closeDelimitter, openDelimitter.length)
  advanceBy(context, openDelimitter.length) // 推进至 {{ 后
  const rawContentLength = closeIndex - openDelimitter.length
  const content = context.source.slice(0, rawContentLength).trim()
  advanceBy(context, rawContentLength + closeDelimitter.length) // 推进至 }} 后

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content
    }
  }
}

function advanceBy(context, index){
  context.source = context.source.slice(index)
}

