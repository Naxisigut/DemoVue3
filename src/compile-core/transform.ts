import { NodeTypes } from "./ast";
import { TO_DISPLAY_STRING } from './runtimeHelpers';

export function transform(root, option = {}){
  const ctx = createContext(root, option)
  tranverseNode(root, ctx)
  
  createCodeGenNode(root) // codegen入口节点
  createCodeGenHelpers(root, ctx)
}

function createCodeGenNode(root: any) {
  root.codeGenNode = root.children[0]
}
function createCodeGenHelpers(root: any, ctx) {
  root.helpers = [...ctx.helpers.keys()]
}

function createContext(root: any, option: any) {
  const ctx = {
    root, 
    NodeTransformers: option.NodeTransformers || [],
    helpers: new Map(),
    help(key){
      ctx.helpers.set(key, 1)
    }
  }

  return ctx
}

function tranverseNode(node, ctx){
  // console.log('node', node);
  const {
    NodeTransformers,
    help
  } = ctx
  for (let i = 0, len = NodeTransformers.length; i < len; i++) {
    const plugin = ctx.NodeTransformers[i];
    plugin(node, ctx)
  }

  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      help(TO_DISPLAY_STRING) // 添加入ctx.helpers内
      break;
    case NodeTypes.ROOT:
    case NodeTypes.ELEMENT:
      tranverseChildren(node, ctx)
      break;
    default:
  }

}

function tranverseChildren(node, ctx){
  const children = node.children
  if(children){ // interpolation 没有children字段
    for (let i = 0; i < children.length; i++) {
      const node = children[i];
      tranverseNode(node, ctx)
    }
  }
}

