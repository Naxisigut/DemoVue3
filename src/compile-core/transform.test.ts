import { expect, describe, it, vi } from 'vitest';
import { transform } from './transform';
import { baseParse } from './parse';
import { NodeTypes } from './ast';
import { generate } from './codegen';

describe('transform', () => {

  it('happy path', () => {
    const ast = baseParse('<div>hi, {{message}}</div>')
    const textPlugin = (node) => {
      if(node.type === NodeTypes.TEXT){
        node.content = 'hi, mini-vue'
      }
      return node
    }
    transform(ast, {
      NodeTransformers: [textPlugin]
    })
  
    const nodeText = ast.children[0].children[0]
    expect(nodeText.content).toBe('hi, mini-vue')
  })
})