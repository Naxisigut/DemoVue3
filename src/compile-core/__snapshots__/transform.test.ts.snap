// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`codegen > happy path 1`] = `
"return function render(_ctx, _cache, $props, $setup, $data, $options) {
    with (_ctx) {
      return \\"hi\\"
    }
  }"
`;
